Try to remove yuni from inifile

guilpier-code opened the pull request:
What was done so far :
- Removal of type **IO::File::Stream**
- Splitting and refactoring ugly function **AnalyzeIniLine(...)**
- Replacing few **Yuni::string** with **std::string** : 
  - **IniFile::filename()** now returns an **std::string**
  - class **IniFile**'s data member **pFilename** is now a **std::string** (and was renamed into **filename_**)
  - Replacement of **AnalyzeIniLine(...)** with some other code removes uses of **Yuni::string** (in args and locally)
- Some more simplifications

Note that the code has reduced  (if we except the addition of new unit tests), and is more clear now.
Please see comments added here to ease the review.

guilpier-code committed:
Remove yuni from inifile : removing Yuni::IO::File::Stream from inifile

guilpier-code committed:
Remove yuni from inifile : tiny simplification

flomnes commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -237,29 +238,20 @@ uint IniFile::Section::size() const
     clear();
-    pFilename = filename;
+    std::string filePath = filename.to<std::string>();
+    const unsigned int lineMaxSize = 2048;
```
```suggestion
```

guilpier-code replied:
Done : declaration of **lineMaxSize** was removed

flomnes commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -237,29 +238,20 @@ uint IniFile::Section::size() const
-    // Load the file
-    IO::File::Stream file;
-    if (file.open(filename))
+    std::ifstream file(filePath);
+    if (file.is_open())
```
Proposal
```cxx
IniFile::readStream(std::istream& in, bool warnings)
{
   // most generic implementation
}
IniFile::open(const std::filesystem::path& file, bool warnings) 
{
   std::ifstream f(path);
   if (f.is_open())
   {
      return readStream(f, warnings);
   }
}
```
So we can have easy in-memory tests with `std::istringstream`

See [this diagram](https://en.cppreference.com/w/cpp/io) for more details.

guilpier-code replied:
I couldn't do exactly as you asked.
Indeed, handling the case where the file could not open correctly cannot easily be done inside the new  **readStream(...)** function.
Not handling this case in the function seems consistent : this function receives a stream and checks its correctness (as you suggested, it's a more generic function).
Handling a file opening error inside this function would have no sense.

guilpier-code replied:
Please note that I made the new function **readStream(...)** **private**.
This makes it difficult to test that function in isolation. And yet, **readStream(...)** is a method called only once, inside a **public** method  (named **open(...)**) of class **iniFile**. So it shouldn't be part of the **iniFile** interface, and should remain **private**.

To overcome this problem, the current class **iniFile** should be split.
For example we coud have a class reading sections from a generic stream instead of file.
And another class could handle the files aspects and use the first one, via its interface.

guilpier-code replied:
**Reminder** : as @sylvlecl said, **readStream(...)** should be made **public** in order to test it, and because it could be useful to have it **public** some day.

guilpier-code replied:
Done

flomnes commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -289,26 +281,25 @@ void IniFile::saveToString(std::string& str) const
     logs.debug() << "  :: writing `" << filename << '`';
-    IO::File::Stream f;
-    if (f.openRW(filename))
+    
+    std::ofstream f(filename.to<std::string>());
+    if (f.good())
```
Same proposal here, don't impose files for output, use `std::ostream`.

guilpier-code replied:
Done

flomnes commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -44,8 +45,8 @@ static inline IniFile::Section* AnalyzeIniLine(const String& pFilename,
     Type type = typeUnknown;
-    char* p = line;
-    const char* key = line;
+    char* p = &line[0];
+    const char* key = &line[0];
```
Horrible

guilpier-code replied:
Yes indeed, but I wanted to touch this function as little as possible, to avoid introducing any side effect.
Any suggestion would be welcome.

This function isn't exactly the most beautiful I've crossed :-)

guilpier-code replied:
**AnalyzeIniLine** disappeared, see comments further

payetvin commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -24,15 +24,16 @@
 namespace Antares
 {
-static inline IniFile::Section* AnalyzeIniLine(const String& pFilename,
+static inline IniFile::Section* AnalyzeIniLine(const std::string filename,
```
use reference or string_view

flomnes replied:
`AnalyzeIniLine` was removed

guilpier-code committed:
Remove yuni from inifile : remove a useless local constant

guilpier-code committed:
Remove yuni from inifile : removing a local use of Yuni::CString

guilpier-code committed:
Remove yuni from inifile : extracting a function reading a stream from a function opening and reading an ini file

guilpier-code committed:
Remove yuni from inifile : when printing ini file, replacing a template out-stream argument with an std::ostream

guilpier-code committed:
Remove yuni from inifile : make a function signature more clear

guilpier-code committed:
Remove yuni from inifile : remove empty lines

guilpier-code committed:
Remove yuni from inifile : changing the type of a private member, from a Yuni string to std::string

guilpier-code committed:
Remove yuni from inifile : refactoring horrible AnalyzeIniLine function + remove useless arg 'warning'

guilpier-code commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -19,99 +19,19 @@
 
 namespace Antares
 {
-static inline IniFile::Section* AnalyzeIniLine(const String& pFilename,
```
Dirty function **AnalyzeIniLine** was removed, and replaced with a clearer block of lines (please see method **readStream(...)**).

guilpier-code commented on src/libs/antares/inifile/include/antares/inifile/inifile.h:
```diff
@@ -310,13 +308,15 @@ class IniFile final
     ** \internal This variable to keep const-correctness for the
     ** public API (cf save).
     */
-    mutable YString pFilename;
```
**pFilename** is now a `std::string`, instead of a `YString`
**pFilename** was renamed into  **filename_**;

guilpier-code committed:
Remove yuni from inifile : correcting a compilation error

guilpier-code committed:
Remove yuni from inifile : file path is removed from args of readStream(...)

It does not make sense for a stream reading function to take file path as an argument.
So using the file path is done at a higher level in the stack

guilpier-code commented on src/libs/antares/benchmarking/file_content.cpp:
```diff
@@ -65,8 +65,6 @@ std::string FileContent::saveToBufferAsIni()
-    std::string buffer;
-    ini.saveToString(buffer);
-    return buffer;
+    return ini.saveToString();
```
method **iniFile::saveToString()** now returns an `std::string`. 
More readable like this.

guilpier-code commented on src/libs/antares/inifile/include/antares/inifile/inifile.h:
```diff
@@ -201,10 +199,10 @@ class IniFile final
     bool save(const AnyString& filename) const;
 
-    void saveToString(std::string& str) const;
+    std::string saveToString() const;
```
As already said, **saveToString()** now returns a `std::string`, instead of taking a reference to a std::string and then changing it inside the function's body. 
More readable like this.

guilpier-code commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -235,44 +153,102 @@ uint IniFile::Section::size() const
+{
+    std::string line;
+    IniFile::Section* currentSection(nullptr);
+    while (std::getline(in_stream, line))
```
Lines inside **while loop** are meant to replace old function **AnalyzeIniLine**.

guilpier-code commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -235,44 +153,102 @@ uint IniFile::Section::size() const
+    return true;
+}
 
-        if (read)
```
To be honest, local variable **read** and call of `Statistics::HasReadFromDisk(...);` were removed.
They were hampering the refactoring and were of no use but for the legacy **GUI**.
As its name suggests, the latter function collects disk loading volume for the **GUI**.

guilpier-code commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -283,29 +259,25 @@ void IniFile::saveToString(std::string& str) const
     if (written != 0)
```
```C++
if (written != 0)
        Statistics::HasWrittenToDisk(written);
```
This could also be removed (see previous comment)

flomnes replied:
Or we could add a callback/virtual class mechanism for specific stuff like this
```cpp
class IEvent
{
public:
    virtual void written(uint size) = 0;
};
saveToString(IEvent& event);
```

guilpier-code replied:
ok, we could do that. 
But it's still more work & tests than a simple removal of : 
```C++
if (written != 0)
        Statistics::HasWrittenToDisk(written);
```
Furthermore, in the **GUI**, printed information about the amount of loaded data may be very approximative.
Besides, these information may also be useless, as the user can have a much more accurate estimation of her/his study's size on disk by asking directly to the system to do it.
By the way, `Statistics::HasWrittenToDisk(written);` increases a ... global variable.

guilpier-code commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -283,29 +259,25 @@ void IniFile::saveToString(std::string& str) const
         for (auto* section = firstSection; section; section = section->next)
             section->saveToStream(f, written);
-
         if (written != 0)
```
`Statistics::HasWrittenToDisk(written);` : same as above : should be removed.

flomnes replied:
I disagree. This is not in the scope of this PR.

guilpier-code replied:
I know, but it's so simple to remove it (please see [the comment just above](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2064#discussion_r1598332129)).

guilpier-code commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -283,29 +259,25 @@ void IniFile::saveToString(std::string& str) const
+    std::ofstream f(filename.to<std::string>());
+    if (f.good())
     {
-        pFilename = filename;
```
`pFilename = filename;` : was removed, as it does not seem to be useful to store the name of a file we're saving. 
And tests on saving studies seem to confirm this uselessness.

flomnes reviewed (changes requested):
As discussed
- `IniFile` -> `IniContent`
- Publish `readStream` and `saveToStream`
- Make tests using these 2 functions above

flomnes reviewed (commented):
What about `Yuni::AnyString` ?

flomnes commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -235,44 +153,102 @@ uint IniFile::Section::size() const
-    clear();
-    pFilename = filename;
+    boost::trim(line);
+    if (boost::starts_with(line, "[") && boost::ends_with(line, "]"))
```
As of C++20, `starts_with` and `ends_with` are [available in the standard library](https://en.cppreference.com/w/cpp/string/basic_string/starts_with) as member functions of `std::string`. Consider using these instead.

guilpier-code replied:
Indeed, a bit more readable like this

guilpier-code replied:
Done

flomnes commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -235,44 +153,102 @@ uint IniFile::Section::size() const
+    return splitLine[1];
+}
+
+bool isProperty(std::string line)
```
```suggestion
static bool isProperty(const std::string& line)
```

guilpier-code replied:
Done

flomnes commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -235,44 +153,102 @@ uint IniFile::Section::size() const
+{
+    if (std::count(line.begin(), line.end(), '=') == 1)
+        return true;
+    return false;
```
```suggestion
    return std::count(line.begin(), line.end(), '=') == 1;
```

flomnes replied:
Also, consider using [`std::ranges::count`](https://en.cppreference.com/w/cpp/algorithm/ranges/count)

guilpier-code replied:
Done

flomnes commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -235,44 +153,102 @@ uint IniFile::Section::size() const
+    std::vector<std::string> splitLine;
+    boost::split(splitLine, line, boost::is_any_of("="));
+
+    return std::make_pair(splitLine[0], splitLine[1]);
```
```suggestion
    return {splitLine[0], splitLine[1]};
```

guilpier-code replied:
I tested your change proposal, but it does not compile, at least on **Windows**. 
It would have been nice though.

flomnes commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -235,44 +153,102 @@ uint IniFile::Section::size() const
+    boost::trim_left(line);
+    if (boost::starts_with(line, "#") || boost::starts_with(line, ";"))
+        return true;
+    return false;
```
1. Use `line.starts_with`
2. One line instead of 3 `return x.starts_with(y) || x.starts_with(z);`

guilpier-code replied:
Done

flomnes commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -235,44 +153,102 @@ uint IniFile::Section::size() const
+
+    if (std::ifstream file(filePath); file.is_open())
+    {
+        if (!readStream(file) && warnings)
+        {
+            logs.error() << "Invalid INI file : " << filePath;
+            return false;
+        }
```
This is very error-prone. If `warning=false`, the function will return `true` even though `readStream` failed.

IMHO `warnings` should decide only on logs, the execution flow shouldn't care about `warnings`. Use nested `if` here.

guilpier-code replied:
ok, you're totally right.
I introduced an extra check to the original code.
In the original code, **warnings** is only used to log when an issue occurs while opening the ini file.
Now it's also used when a format issue occurs.
So I inadvertently changed the behavior.

So, in order to revert this new unsuitable behavior, the statement : 
`if (!readStream(file) && warnings)` 
should be replaced with : 
`if (!readStream(file))`

flomnes commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -235,44 +153,102 @@ uint IniFile::Section::size() const
 }
 
-void IniFile::saveToString(std::string& str) const
+std::string IniFile::saveToString() const
```
```suggestion
std::string IniFile::toString() const
```

guilpier-code replied:
Done

flomnes committed:
Restore call to Statistics::HasReadFromDisk

sylvlecl committed:
Merge remote-tracking branch 'origin/develop' into fix/remove_yuni_from_inifile

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Fix missing dependency to boost for ini file

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

guilpier-code commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -253,51 +168,119 @@ uint IniFile::Section::size() const
+        if (isProperty(line))
+        {
+            std::pair<std::string, std::string> pair = getKeyValuePair(line);
+            currentSection->add(pair.first, pair.second);
```
`currentSection->add(pair.first, pair.second);` : what if we're reading a (key, value) pair and we're not in a section (meaning : **currentSection** is still **nullptr**)  ?
We should raise an error in this case.

guilpier-code replied:
Done

guilpier-code committed:
Remove yuni from inifile : corrections due to review (see commit message)

- make little function about reading the ini file static and more simple (less code lines).
- avoid use of boost functions when STL offers the same service
- replace std::count with std::ranges::count

guilpier-code commented:
> What about `Yuni::AnyString` ?

Please see the PR's presentation comment (now more detailed)

guilpier-code committed:
Remove yuni from inifile : removing a behavior difference with original code when handling issues in reading ini files

guilpier-code committed:
Remove yuni from inifile : renaming only (IniFile::saveToString() ==> IniFile::toString())

guilpier-code committed:
Remove yuni from inifile : when reading a ini file, handling case when a property is not in a section

guilpier-code commented:
> As discussed
> 
> * `IniFile` -> `IniContent`
> * Publish `readStream` and `saveToStream`
> * Make tests using these 2 functions above

I'm not sure that renaming class **IniFile** into **IniContent** makes sense for now, because it is still coupled to the **file** concept : 
For example :
- it still contains a data member **filename_**
- it still has functions like **filename()** or **loaded()**

We should properly decouple data reading/saving from content managing.

guilpier-code committed:
Remove yuni from inifile : make function readStream a public function for tests reasons

guilpier-code committed:
Remove yuni from inifile : few simplifications

guilpier-code committed:
Remove yuni from inifile : oops, forgot a ";" somewhere

guilpier-code committed:
Remove yuni from inifile : few simplifications

guilpier-code committed:
Remove yuni from inifile : restore old behavior : a property not in a section is skipped

guilpier-code committed:
Remove yuni from inifile : small simplification

sylvlecl commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -253,51 +160,110 @@ uint IniFile::Section::size() const
+    return splitLine[1];
+}
+
+static bool isProperty(std::string line)
```
either use a const ref or a string_view

guilpier-code replied:
Done

sylvlecl commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -253,51 +160,110 @@ uint IniFile::Section::size() const
-    // Load the file
-    IO::File::Stream file;
-    if (file.open(filename))
+static std::string getSectionName(std::string line)
```
either use a const ref or a string_view when possible to avoid copies

guilpier-code replied:
Done

sylvlecl commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -253,51 +160,110 @@ uint IniFile::Section::size() const
+    return std::ranges::count(line.begin(), line.end(), '=') == 1;
+}
+
+static std::pair<std::string, std::string> getKeyValuePair(std::string line)
```
Note that you could create a small struct for naming key and value (only 4 lines of code), instead of using a std::pair<std::string, std::string> which is less expressive.

sylvlecl replied:
Or juste return a `Property` (existing class).

guilpier-code replied:
I like your second idea, done.

sylvlecl commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -253,51 +160,110 @@ uint IniFile::Section::size() const
     {
-        enum
+        read += line.size();
+        if (startingSection(line))
```
isStartingSection

guilpier-code replied:
Done

sylvlecl commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -253,51 +160,110 @@ uint IniFile::Section::size() const
-        uint64_t read = 0;
+        if (isProperty(line))
+        {
+            // Note : if a property not in a section, it's simply skipped
```
Why not get an error instead?
Was it the previous behaviour ?

guilpier-code replied:
I first raised an error in the code when a property was not inside a section.
But I realized that the previous behavior was to skip such properties.
That's why it's emphasized in the code as a comment.

sylvlecl reviewed (dismissed):
The implementation is indeed MUCH simpler, that's nice !

However we need to add unit tests to cover this feature and ensure there is no regression in edge cases, like comments, properties which are not in a section, and so on.

Answer from @guilpier-code : it's now done

guilpier-code committed:
Remove yuni from inifile : several simplification (see commit message)

- Avoiding code duplication by defining method IniFile::saveToStream
- Renaming argument "file" into "stream_out", more matching its type
- Simplifying constructors definitions for class Property
- Introducing "#pragma once" instead of old macros
-

flomnes commented on src/libs/antares/inifile/include/antares/inifile/inifile.h:
```diff
@@ -68,7 +66,7 @@ class IniFile final
-        Section();
+        Section() = default;
         explicit Section(const AnyString& name);
         ~Section();
```
`= default` ?

guilpier-code replied:
yes, why not ?

flomnes replied:
It was not `default` previously

guilpier-code committed:
Remove yuni from inifile : adding few tests on reading a stream to populate a Inifile object

guilpier-code committed:
Remove yuni from inifile : adding more tests on reading a stream to populate a Inifile object

guilpier-code committed:
Remove yuni from inifile : adding tests for saving iniFile to a stream + simplifications

guilpier-code committed:
Remove yuni from inifile : corrections due to review

guilpier-code commented:
I can see 3 possible directions for improving the class **IniFile** : 
- **Yuni** : remove some more **Yuni** things (mainly strings)
- **Implementation details** : several things can be done to improve, let's say, the implementation details. I mean without changing current classes interfaces. These improvements could make the code shrink quite a lot. For example : 
  - classes **IniFile** or **Section** internally implement a (chained) **list**. For both of them, a new **list** mechanism is duplicated. It's crazy :-). So let's replace it with a container from the **STL**. It could be a **std::list**, but a **std::vector** would work the same I guess. 
  - We have **each(...)** that are barley used : instead, **IniFile** and **Section** are traversed using classic **for loops**, even internally.
- **Design** :  as said above [here](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2064#issuecomment-2108117280), **IniFile** has drawbacks that would require to split it.

guilpier-code committed:
Remove yuni from inifile : adding more tests about saving ini content

flomnes commented on src/libs/antares/inifile/inifile.cpp:
```diff
@@ -253,95 +147,143 @@ uint IniFile::Section::size() const
+            // Note : if a property not in a section, it's simply skipped
+            if (currentSection)
+            {
+                IniFile::Property p = getProperty(line);
+                currentSection->add(p.key, p.value);
```
```suggestion
                currentSection->add(Property(line));
```
Add a constructor for `Property` : `Property(std::string line)`

guilpier-code replied:
If we do that, we couple **reading concern** (a line) and **business logic** (Property).
As it's done now, **getProperty** is in the **reading** area, and depend on type **Property** (business).
If we introduce **Property(std::string line)** **reading** puts a step into business logic (dependency in the other direction).

guilpier-code replied:
If we want to merge those 2 lines in one, we could have a new **Section::add(...)** function, taking a Property as argument, so we could have : 
`currentSection->add(getProperty(line));`

guilpier-code replied:
Done

flomnes commented on src/tests/src/libs/antares/inifile/test_inifile_io.cpp:
You should add a test or 2 on **keys** when reading INI files. Indeed, you look for sections & properties, but never check the actual content of properties.

guilpier-code replied:
Done

guilpier-code committed:
Remove yuni from inifile : more tests on reading ini content

guilpier-code committed:
[skip ci] Remove yuni from inifile : just renaming a test

guilpier-code committed:
[skip ci] Remove yuni from inifile : oops

guilpier-code committed:
Remove yuni from inifile : adding a test on created properties content when reading a ini content

sylvlecl commented on src/tests/src/libs/antares/inifile/test_inifile_io.cpp:
```diff
@@ -0,0 +1,233 @@
+
```
Missing header

sylvlecl commented on src/tests/src/libs/antares/inifile/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+
+set(src_libs_antares "${CMAKE_SOURCE_DIR}/libs/antares")
```
Not needed

sylvlecl reviewed (changes requested):
Last details

sylvlecl committed:
Missing copyright header

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Unnecessary line

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Formatting

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl reviewed (approved):
Just applied copyright + formatting
